paddle/fluid/operators/controlflow/recurrent_op_helper.cc

```diff
@@ -0,0 +1,264 @@
+    const std::vector<std::string> &var_names) {
+  std::vector<std::string> retv;
+  retv.reserve(var_names.size());
+  std::transform(var_names.begin(), var_names.end(), std::back_inserter(retv),
+                 framework::GradVarName);
+  return retv;
+}
+
+// Set memory vars in recurrent op as skip vars.
+static void SetOpMemVarsAsSkip(const OpVariant &op, bool set_grad) {
+  bool has_state = op.Attr<bool>(kHasStates);
+  if (has_state) {
+    std::unordered_set<std::string> skip_vars;
+
+    auto &mem_vars = op.Attr<std::vector<std::string>>(kStates);
+    skip_vars.insert(mem_vars.begin(), mem_vars.end());
+
+    auto &pre_mem_vars = op.Attr<std::vector<std::string>>(kExStates);
+    skip_vars.insert(pre_mem_vars.begin(), pre_mem_vars.end());
+
+    if (set_grad) {
+      auto mem_grad_vars = GradVarLists(mem_vars);
+      skip_vars.insert(mem_grad_vars.begin(), mem_grad_vars.end());
+      auto pre_mem_grad_vars = GradVarLists(pre_mem_vars);
+      skip_vars.insert(pre_mem_grad_vars.begin(), pre_mem_grad_vars.end());
+    }
+    SetSkipVars(op, skip_vars);
+  }
+}
+
+// Set outputs and memory vars of the input forward op as skip vars
+static void SetRecurrentForwardOpOnlySkipVarAttr(const OpVariant &fwd_op) {
+  SetOpMemVarsAsSkip(fwd_op, false);
+
+  std::unordered_set<std::string> fwd_skip_vars;
+  auto &output_vars = fwd_op.Outputs().at(kOutputs);
+  for (const std::string &name : output_vars) {
+    fwd_skip_vars.insert(name);
+  }
+  SetSkipVars(fwd_op, fwd_skip_vars);
+}
+
+// Set skip vars of matched recurrent op and recurrent_grad op
+static void SetRecurrentOpAndRecurrentGradOpSkipVarAttr(
+    const OpVariant &fwd_op, const OpVariant &bwd_op) {
+  // Find all skippable variables in forward recurrent_op
+  SetOpMemVarsAsSkip(fwd_op, false);
+
+  auto *grad_block = bwd_op.Attr<framework::BlockDesc *>(kStepBlock);
+  std::unordered_set<std::string> fwd_skip_vars;
+  for (auto *op_desc : grad_block->AllOps()) {
+    for (auto &in_arg_name : op_desc->InputArgumentNames()) {
+      if (IsSkippableVar(in_arg_name, grad_block)) {
+        fwd_skip_vars.insert(in_arg_name);
+      }
+    }
+    for (auto &out_arg_name : op_desc->OutputArgumentNames()) {
+      if (IsSkippableVar(out_arg_name, grad_block)) {
+        fwd_skip_vars.insert(out_arg_name);
+      }
+    }
+  }
+  SetSkipVars(fwd_op, fwd_skip_vars);
+
+  // Find all skippable variables in recurrent_grad_op
+  // The skippable variables are those which would be used across time steps
+  SetOpMemVarsAsSkip(bwd_op, true);
+  std::unordered_set<std::string> bwd_skip_vars;
+
+  auto &fwd_input = fwd_op.Inputs().at(kInputs);
+  auto &in_grads = bwd_op.Outputs().at(framework::GradVarName(kInputs));
+  // auto &param_grads =
+  // bwd_op.Outputs().at(framework::GradVarName(kParameters));
+
+  PADDLE_ENFORCE_EQ(
+      fwd_input.size(), in_grads.size(),
+      "Backward input gradient number does not match forward input number.");
+  for (size_t i = 0; i < in_grads.size(); ++i) {
```

Comment:
Seems wrong. You should review the codes of [`recurrent_op.cc`](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/recurrent_op.cc#L448).

Reply:
Discussed offline, it may not be wrong.

Reply:
Discussed offline, it may not be wrong.